### PR TITLE
Time Zone Support

### PR DESCRIPTION
**Pull request by aaron-siegel**

### CODE REVIEW

**Comment by aaron-siegel** [comment]
This change is [Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/250)

**Review by mkornacker — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/250)* status: 0 of 18 files reviewed, 11 unresolved discussions (waiting on @aaron-siegel)

___
*[`pixeltable/env.py` line 109 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6S5HSK-Wu3bz5X8IJc:-O6S5HSLBvJswNYCfG-S:b4xu71c) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/env.py#L109)):*
> ```Python
>     @property
>     def default_time_zone(self) -> Optional[ZoneInfo]:
>         return self._default_time_zone
> ```

This is missing in init.

It would be even better to move the declarations in init to the class level.
___
*[`pixeltable/env.py` line 122 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6S79OcEgQ60kptHjKv:-O6S79Od02W27Na_ol95:b-auyn23) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/env.py#L122)):*
> ```Python
>             # name, using the latter to set an explicit default time zone in Python.
>             with self.engine.begin() as conn:
>                 conn.execute(sql.text('SET TIME ZONE LOCAL;'))
> ```

Does it need a ';'? I wouldn't expect that.
___
*[`pixeltable/env.py` line 123 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6S6iLo3Ff-UGcu7yd6:-O6S6iLo3Ff-UGcu7yd7:bcxwjsu) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/env.py#L123)):*
> ```Python
>             with self.engine.begin() as conn:
>                 conn.execute(sql.text('SET TIME ZONE LOCAL;'))
>             with self.engine.begin() as conn:
> ```

why two separate transactions?
___
*[`pixeltable/exprs/literal.py` line 27 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SBrrU2WZ-F6NIOl5z:-O6SBrrU2WZ-F6NIOl6-:bxgazsh) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/exprs/literal.py#L27)):*
> ```Python
>         super().__init__(col_type)
>         if isinstance(val, datetime.datetime):
>             # Normalize the datetime to UTC: all timestamps are stored as UTC (both in the database and in literals)
> ```

Doesn't Postgres already do this for us?
___
*[`pixeltable/metadata/converters/convert_19.py` line 20 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SERcK6QKGnEjpq9BV:-O6SERcK6QKGnEjpq9BW:b-cyogib) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/metadata/converters/convert_19.py#L20)):*
> ```Python
>     with engine.begin() as conn:
>         for tbl_version in Catalog.get().tbl_versions.values():
>             for col in tbl_version.cols:
> ```

This isn't sufficient, you also need to alter the index columns.
___
*[`pixeltable/metadata/converters/convert_19.py` line 39 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SEw2PC24UKEasxYOq:-O6SEw2PC24UKEasxYOr:b84afox) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/metadata/converters/convert_19.py#L39)):*
> ```Python
>         # timestamp literal, which (in version 19) is stored in the DB as a naive datetime.
>         # We convert it to an aware datetime, stored in UTC.
>         assert v['val_t'] == pxt.ColumnType.Type.TIMESTAMP.name
> ```

Also assert for _classname, just for good measure
___
*[`tests/functions/test_timestamp.py` line 77 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SGeEO1TtsyD2BiiIU:-O6SGeEO1TtsyD2BiiIV:b-fcayqu) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L77)):*
> ```Python
>             # an IANA zone, but remain the same if it's expressed as a hardcoded offset. Thus `replace()` will give
>             # different answers in each case. None of the other functions depend on this distinction.
>             # TODO: We might consider explicitly casting timestamps to the IANA time zone in SqlScanNode, as soon as
> ```

That was also my first thought when I read the preceding explanation.

You don't have to look at every value, only the ones for timestamp-typed expressions. I think this would be a relatively cheap per-row postprocessing step in SqlNode.__next__()
___
*[`tests/functions/test_timestamp.py` line 101 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SHfRj9TxDNV1TBBPC:-O6SHfRj9TxDNV1TBBPD:b-jjvwai) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L101)):*
> ```Python
>         timestamps = [
>             datetime.fromisoformat('2024-07-01T22:45:12'),
>             datetime.fromisoformat('2024-07-01T22:45:12-02:00'),
> ```

explanation: why these specifically?
___
*[`tests/functions/test_timestamp.py` line 107 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SIQmT2rOwsKi-4Vhh:-O6SIQmT2rOwsKi-4Vhi:b-b9fhrv) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L107)):*
> ```Python
>         props_to_test = ['year', 'month', 'day', 'hour', 'minute']
>         methods_to_test = ['weekday', 'isoweekday']
>         query_time_zone = ZoneInfo('Europe/London')
> ```

Europe/Berlin :)
___
*[`tests/functions/test_timestamp.py` line 130 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SIsYr32uyruqmXu5g:-O6SIsYr32uyruqmXu5h:b-yidrhy) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L130)):*
> ```Python
> 
>             assert len(results) == len(timestamps)
>             for row in range(len(timestamps)):
> ```

Can you use row_idx or something like that here? 'row' always means 'row containing values' elsewhere.
___
*[`tests/io/test_pandas.py` line 99 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SFyJr6REpFuf6cvPr:-O6SFyJr6REpFuf6cvPs:blvazb4) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/io/test_pandas.py#L99)):*
> ```Python
>         assert result_set['string'] == ['fish', 'cake', 'salad', 'egg']
>         assert result_set['string_n'] == ['fish', 'cake', None, 'egg']
>         assert result_set['ts'] == [datetime.datetime(2024, 5, n).astimezone(None) for n in range(3, 7)]
> ```

Why does this require astimezone()?

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/250)* status: 0 of 19 files reviewed, 11 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 109 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6S5HSK-Wu3bz5X8IJc:-O6XBLuO4lZEIuWMmJYv:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/env.py#L109)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

This is missing in init.

It would be even better to move the declarations in init to the class level.
</blockquote></details>

Done.
___
*[`pixeltable/env.py` line 122 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6S79OcEgQ60kptHjKv:-O6XCaDzEfSzZndeZIQK:bmhw22x) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/env.py#L122)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Does it need a ';'? I wouldn't expect that.
</blockquote></details>

You're right, it doesn't. Fixed.
___
*[`pixeltable/env.py` line 123 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6S6iLo3Ff-UGcu7yd6:-O6XCbhp-Zj-U7NqJFdd:b-z7h8hg) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/env.py#L123)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

why two separate transactions?
</blockquote></details>

If they're bundled as a single transaction, the change doesn't always get picked up properly by the subsequent SHOW TIME ZONE. i don't know why this is (possibly a bug in postgres?)
___
*[`pixeltable/exprs/literal.py` line 27 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SBrrU2WZ-F6NIOl5z:-O6XDDTN02FF6xtM4sLK:bf6od0w) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/exprs/literal.py#L27)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Doesn't Postgres already do this for us?
</blockquote></details>

Not if we're storing it as a literal inside a JSON (metadata) struct.
___
*[`pixeltable/metadata/converters/convert_19.py` line 20 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SERcK6QKGnEjpq9BV:-O6XKATp3zjKoXMeXctB:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/metadata/converters/convert_19.py#L20)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

This isn't sufficient, you also need to alter the index columns.
</blockquote></details>

Done.
___
*[`pixeltable/metadata/converters/convert_19.py` line 39 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SEw2PC24UKEasxYOq:-O6XEE7k14BG2Y7-1QHC:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/metadata/converters/convert_19.py#L39)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Also assert for _classname, just for good measure
</blockquote></details>

Done.
___
*[`tests/functions/test_timestamp.py` line 101 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SHfRj9TxDNV1TBBPC:-O6XKXl59cDBR_Z0CZQI:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L101)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

explanation: why these specifically?
</blockquote></details>

Done.
___
*[`tests/functions/test_timestamp.py` line 107 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SIQmT2rOwsKi-4Vhh:-O6XHALW8-XvLsEDghl2:b-m2nrlc) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L107)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Europe/Berlin :)
</blockquote></details>

Fine :)
___
*[`tests/functions/test_timestamp.py` line 130 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SIsYr32uyruqmXu5g:-O6XKe4nABjBNNSb-sK-:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/functions/test_timestamp.py#L130)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Can you use row_idx or something like that here? 'row' always means 'row containing values' elsewhere.
</blockquote></details>

Done.
___
*[`tests/io/test_pandas.py` line 99 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SFyJr6REpFuf6cvPr:-O6XL5Y46Nw8jIytBvDL:btm9u66) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/tests/io/test_pandas.py#L99)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Why does this require astimezone()?
</blockquote></details>

All timestamps that come out of the database are now aware, regardless of whether they were aware when they were inserted. And Python will complain if you try to compare a naive datetime to an aware one. This makes it into an aware datetime (with system timezone, which is the default for unit tests that don't explicitly set it to something else)

I added a comment explaining this

File: pixeltable/env.py [hunk 1]
```diff
@@ -299,46 +312,57 @@ def _set_up(self, echo: bool = False, reinit_db: bool = False) -> None:
...
+            else:
+                try:
+                    _ = ZoneInfo(tz_name)
+                except ZoneInfoNotFoundError:
+                    self._logger.error(f'Invalid time zone specified in configuration: {tz_name}')
+
+        if reinit_db and self._store_db_exists():
+            self._drop_store_db()
+
+        should_create_db = not self._store_db_exists()
```
Comment (mkornacker) [inline pixeltable/env.py]: 'should' as opposed to just 'create_db'? (for most people, 'should' does not imply 'will' :))

**Review by mkornacker — APPROVED** [review]

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/250)* status: 0 of 19 files reviewed, 12 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/env.py` line 329 at r4](https://reviewable.io/reviews/pixeltable/pixeltable/250#gh-1761884596:-O6wr_DF0o-AYedeybTD:b-4q9i2d) ([raw file](https://github.com/pixeltable/pixeltable/blob/3a2cce8242aa7126a1580670bd91ec41ce855c55/pixeltable/env.py#L329)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

'should' as opposed to just 'create_db'? (for most people, 'should' does not imply 'will' :))
</blockquote></details>

changed to `create_db`
___
*[`pixeltable/metadata/converters/convert_19.py` line 20 at r2](https://reviewable.io/reviews/pixeltable/pixeltable/250#-O6SERcK6QKGnEjpq9BV:-O6XcaZv6bjbQ3sNG9XU:bqyhepb) ([raw file](https://github.com/pixeltable/pixeltable/blob/60eeaa5148750021815a92a62e4e4b4dd6615d7d/pixeltable/metadata/converters/convert_19.py#L20)):*
<details><summary><i>Previously, aaron-siegel (Aaron Siegel) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Per conversation today, index columns are already taken care of since they're included in `tbl_version.cols`. So no special handling of index columns is needed.

**Pull request merged**